internal/controllers/configmaps.go

```diff
@@ -141,17 +145,201 @@ func (r *Reconciler) reconcileOAuth2ProxyConfig(ctx context.Context, cr *model.C
+	tcp_nopush          on;
+	keepalive_timeout   65;
+	types_hash_max_size 4096;
+
+	include             /etc/nginx/mime.types;
+	default_type        application/octet-stream;
+
+	server {
+		server_name {{ .ServerName }};
+
+		{{ if .TLSEnabled -}}
+		listen {{ .ContainerPort }} ssl;
+		listen [::]:{{ .ContainerPort }} ssl;
+
+		ssl_certificate {{ .TLSCertFile }};
+		ssl_certificate_key {{ .TLSKeyFile }};
+
+		ssl_session_timeout 5m;
+		ssl_session_cache shared:SSL:20m;
+		ssl_session_tickets off;
+
+		ssl_dhparam {{ .DHParamFile }};
+
+		# intermediate configuration
+		ssl_protocols TLSv1.2 TLSv1.3;
+		ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305;
+		ssl_prefer_server_ciphers off;
+
+		# HSTS (ngx_http_headers_module is required) (63072000 seconds)
+		add_header Strict-Transport-Security "max-age=63072000" always;
+
+		# OCSP stapling
+		ssl_stapling on;
+		ssl_stapling_verify on;
+
+		ssl_trusted_certificate {{ .CACertFile }};
+
+		# Client certificate authentication
+		ssl_client_certificate {{ .CACertFile }};
+		ssl_verify_client on;
+
+		{{- else -}}
+
+		listen {{ .ContainerPort }};
+		listen [::]:{{ .ContainerPort }};
+
+		{{- end }}
+
+		{{ range .AllowedPathPrefixes -}}
+		location {{ . }} {
+			proxy_pass http://127.0.0.1:{{ $.CryostatPort }}$request_uri;
+		}
+
+		{{ end -}}
+
+		location / {
+			return 404;
+		}
+	}
+
+	# Heatlh Check
+	server {
+		listen {{ .HealthPort }};
+		listen [::]:{{ .HealthPort }};
+
+		location = /healthz {
+			return 200;
+		}
+
+		location / {
+			return 404;
+		}
+	}
+}`))
+
+const (
+	dhFileName = "dhparam.pem"
+	// From https://ssl-config.mozilla.org/ffdhe2048.txt
+	dhParams = `-----BEGIN DH PARAMETERS-----
+MIIBCAKCAQEA//////////+t+FRYortKmq/cViAnPTzx2LnFg84tNpWp4TZBFGQz
++8yTnc4kmz75fS/jY2MMddj2gbICrsRhetPfHtXV/WVhJDP1H18GbtCFY2VVPe0a
+87VXE15/V8k1mE8McODmi3fipona8+/och3xWKE2rec1MKzKT0g6eXq8CrGCsyT7
+YdEIqUuyyOP7uWrat2DX9GgdT0Kj3jlN9K5W7edjcrsZCwenyO4KbXCeAvzhzffi
+7MA0BM0oNC9hkXL+nOmFg/+OTxIy7vKBg8P+OxtMb61zO7X8vC7CIAXFjvGDfRaD
+ssbzSibBsu/6iGtCOGEoXJf//////////wIBAg==
+-----END DH PARAMETERS-----`
+)
+
+func (r *Reconciler) reconcileAgentProxyConfig(ctx context.Context, cr *model.CryostatInstance, tls *resources.TLSConfig) error {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name + "-agent-proxy",
+			Namespace: cr.InstallNamespace,
+		},
+	}
+
+	data := map[string]string{}
+	buf := &bytes.Buffer{}
+	params := &nginxConfParams{
+		ServerName:    fmt.Sprintf("%s-agent.%s.svc", cr.Name, cr.InstallNamespace),
+		ContainerPort: constants.AgentProxyContainerPort,
+		HealthPort:    constants.AgentProxyHealthPort,
+		CryostatPort:  constants.CryostatHTTPContainerPort,
+		AllowedPathPrefixes: []string{
+			"/api/v2.2/discovery/",
+			"/api/v2.2/credentials/",
```

Comment:
These paths will need to be updated at some point, depending on the order in which these PRs are merged:

https://github.com/cryostatio/cryostat/pull/584

I wonder if there's anything that can be done to make this "interface" (API paths subset) more stable. My initial thought was for the Cryostat server to expose some paths with a common prefix like `*/api/agent*/vx/foo` so that the proxy could just allow this one prefix, but then the server needs to duplicate the functionality of each of these endpoints under that new path prefix. Or else, the server returns redirection responses, but then the proxy still needs to handle these path prefixes where the client (agent) will get redirected to.

I have some other vague ideas around JWT tokens and things but I think it just ends up more complex and fragile than it needs to be...

Reply:
There is actually a fairly nice Cryostat server side solution here:

https://dennis-xlc.gitbooks.io/restful-java-with-jax-rs-2-0-2rd-edition/content/en/part1/chapter4/path.html

The path parameter expression syntax could be useful, so that instead of just `@Path("/api/v4/credentials")`, it could be `@Path("/api/{a:v4|agent}/credentials")`, which would mean the endpoint should be equally available with either `/api/v4` or `/api/agent` prefix. Then we can clean this up so that the TLS proxy only passes through requests with the `/api/agent` prefix, and refactor the Agent HTTP client to use the endpoints with that prefix.

Then if we add or remove some functionality in the future that changes the endpoints that Cryostat and the Agent use to communicate with each other, we don't also need to update the TLS proxy.

Reply:
Very cool! Should I update the path filter now? Or in a follow up PR?